crypto/encrypt_test.go

```diff
@@ -141,3 +141,69 @@ func TestTwofishEntropy(t *testing.T) {
 		t.Error("supposedly high entropy ciphertext has been compressed!")
 	}
 }
+
+// TestUnitCiphertextUnmarshalJSON tests that Ciphertext.UnmarshalJSON
+// correctly fails on invalid JSON marshalled Ciphertext.
+func TestUnitCiphertextUnmarshalJSON(t *testing.T) {
+	// Test unmarshalling invalid JSON.
+	invalidJSONBytes := [][]byte{
+		nil,
+		[]byte{},
+		[]byte("\""),
+	}
+	for _, jsonBytes := range invalidJSONBytes {
+		var ct Ciphertext
+		err := ct.UnmarshalJSON(jsonBytes)
+		if err == nil {
+			t.Errorf("expected unmarshall to fail on the invalid JSON: %q\n", jsonBytes)
+		}
+	}
+}
+
+// TestCiphertextMarshalling tests that marshalling Ciphertexts to JSON results
+// in the expected JSON. Also tests that marshalling that JSON back to
+// Ciphertext results in the original Ciphertext.
+func TestCiphertextMarshalling(t *testing.T) {
+	// Ciphertexts and corresponding JSONs to test marshalling and unmarshalling.
+	ciphertextMarshallingTests := []struct {
+		ct        Ciphertext
+		jsonBytes []byte
+	}{
+		{ct: Ciphertext(nil), jsonBytes: []byte("null")},
+		{ct: Ciphertext(""), jsonBytes: []byte(`""`)},
+		{ct: Ciphertext("a ciphertext"), jsonBytes: []byte(`"YSBjaXBoZXJ0ZXh0"`) /* base64 encoding of the Ciphertext */},
+	}
+	for _, test := range ciphertextMarshallingTests {
+		expectedCt := test.ct
+		expectedJSONBytes := test.jsonBytes
+
+		// Create a copy of expectedCt so Unmarshalling does not modify it, as
+		// we need it later for comparison.
+		var ct Ciphertext
+		if expectedCt == nil {
+			ct = nil
+		} else {
+			ct = make(Ciphertext, len(expectedCt))
+			copy(ct, expectedCt)
+		}
+
+		// Marshal ct to JSON.
+		jsonBytes, err := ct.MarshalJSON()
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !bytes.Equal(jsonBytes, expectedJSONBytes) {
+			t.Fatalf("Ciphertext %#v marshalled incorrectly: expected %q, got %q\n", ct, expectedJSONBytes, jsonBytes)
```

Comment:
If %v is used then `Ciphertext(nil)` and `Ciphertext([]byte{})` will be printed identically as "Ciphertext [] ". Using %#v prints "crypto.Ciphertext(nil)" and "crypto.Ciphertext{}", respectively.